// Copyright 2019 The Google Research Authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Protocol buffer specification for rewrite rules of the morphotactics FST.

syntax = "proto2";

package analyzer.src.morphotactics;


// Internal representation of finite-state transducer rewrite rules used in
// compiling morphotactics FST. Lexicon entries (from TSV files under
// //analyzer/src/lexicon/base) and morphotactics model (from txt files under
// //analyzer/src/morphotactics/model) are validated and parsed into
// RewriteRule objects throughout text FST compilation in
// //analyzer/src/morphotactics/model_compile.py. This is basically the
// structured representation of the AT&T FSM format.
message RewriteRule {
  optional string from_state = 1;  // source state.
  optional string to_state = 2;  // destination state.
  optional string input = 3;  // arc label on input tape.
  optional string output = 4;  // arc label on output tape.
}

// Ad hoc container of rewrite rules. Mainly used to conveniently consume
// serialized arrays of rewrite rules in unit tests of //analyzer/src/lexicon
// and //analyzer/src/morphotactics submodules.
message RewriteRuleSet {
  repeated RewriteRule rule = 1;
}
